theano/tensor/nnet/tests/test_corr.py

```diff
@@ -0,0 +1,313 @@
+from nose.plugins.skip import SkipTest
+from nose.plugins.attrib import attr
+import numpy
+
+import theano
+import theano.tensor as T
+from theano.tests import unittest_tools as utt
+from theano.tensor.nnet import corr, conv
+from theano.tensor.basic import _allclose
+
+
+class TestCorr2D(utt.InferShapeTester):
+    if theano.config.mode == "FAST_COMPILE":
+        mode = theano.compile.get_mode("FAST_RUN")
+    else:
+        mode = None
+    dtype = theano.config.floatX
+
+    def setUp(self):
+        super(TestCorr2D, self).setUp()
+        self.input = T.tensor4('input', dtype=self.dtype)
+        self.input.name = 'default_V'
+        self.filters = T.tensor4('filters', dtype=self.dtype)
+        self.filters.name = 'default_filters'
+        if not conv.imported_scipy_signal and theano.config.cxx == "":
+            raise SkipTest("conv2d tests need SciPy or a c++ compiler")
+
+    def validate(self, image_shape, filter_shape,
+                 border_mode='valid', subsample=(1, 1),
+                 input=None, filters=None,
+                 verify_grad=True, should_raise=False,
+                 non_contiguous=False):
+        """
+        :param image_shape: The constant shape info passed to corrMM.
+        :param filter_shape: The constant shape info passed to corrMM.
+        """
+        N_image_shape = [T.get_scalar_constant_value(T.as_tensor_variable(x))
+                         for x in image_shape]
+        N_filter_shape = [T.get_scalar_constant_value(T.as_tensor_variable(x))
+                          for x in filter_shape]
+
+        if input is None:
+            input = self.input
+        if filters is None:
+            filters = self.filters
+
+        # THEANO IMPLEMENTATION
+
+        # we create a symbolic function so that verify_grad can work
+        def sym_CorrMM(input, filters):
+            # define theano graph and function
+            input.name = 'input'
+            filters.name = 'filters'
+            rval = corr.CorrMM(border_mode, subsample)(input, filters)
+            rval.name = 'corr_output'
+            return rval
+
+        output = sym_CorrMM(input, filters)
+        output.name = 'CorrMM()(%s,%s)' % (input.name, filters.name)
+        theano_corr = theano.function([input, filters], output, mode=self.mode)
+
+        # initialize input and compute result
+        image_data = numpy.random.random(N_image_shape).astype(self.dtype)
+        filter_data = numpy.random.random(N_filter_shape).astype(self.dtype)
+        if non_contiguous:
```

Comment:
@nouiz, this is my attempted at testing the non-contiguous case. Let me know if there is a better way. Added new test that uses this flag.